### PR TITLE
Forward-looking (FL) and Detailed Balance (DB) losses

### PR DESCRIPTION
**Pull request by AlexandraVolokhova**
Forward-looking loss implementation based on  https://arxiv.org/abs/2302.01687 and the previous PR: https://github.com/alexhernandezgarcia/gflownet/pull/127

Detailed balance loss implementation based on https://arxiv.org/abs/2201.13259

Wandb report with comparison FL v.s. TB on ctorus and grid: https://wandb.ai/alexandra-volokhova/gfn-dev/reports/Forward-looking-loss-dev--Vmlldzo2MTE4MzMw 

Added sanity check experiments with new DB and FL losses (see [wandb](https://wandb.ai/alexhg/gfn_sanity_checks))

### CODE REVIEW

File: gflownet/gflownet.py [hunk 1]
```diff
@@ -663,6 +702,79 @@ def trajectorybalance_loss(self, it, batch):
...
+        )
+        masks_f = batch.get_masks_forward(of_parents=True)
+        policy_output_f = self.forward_policy(parents_policy)
+        logprobs_fwd = self.env.get_logprobs(
+            policy_output_f, actions, masks_f, parents, is_backward=False
+        )
+
+        states_log_flflow = self.state_flow(states_policy)
+        # forward-looking flow is 1 in the terminal states
+        states_log_flflow[done.eq(1)] = 0.0
```
Comment (carriepl) [inline gflownet/gflownet.py:753]: Here the comment says that it should be set to 1 (and I also think it should be 1) but the code is setting it to 0.

File: gflownet/gflownet.py [hunk 2]
```diff
@@ -663,6 +702,79 @@ def trajectorybalance_loss(self, it, batch):
...
+        # Can be optimised by reusing states_log_flflow and batch.get_parent_indices
+        parents_log_flflow = self.state_flow(parents_policy)
+
+        assert batch.non_terminal_rewards
+        rewards_states = batch.get_rewards()
+        rewards_parents = batch.get_rewards_parents()
+        energies_states = -torch.log(rewards_states)
+        energies_parents = -torch.log(rewards_parents)
```
Comment (carriepl) [inline gflownet/gflownet.py:761]: At least for testing purposes, it would be useful to implement the feature where null rewards are changed to 1 (which, by making their energies = 0, will collapse the FL loss to the DB loss when no rewards are computed for non-terminal states). This should allow you to debug in a simpler setup before trying to make the full FL loss work.

File: gflownet/gflownet.py [hunk 3]
```diff
@@ -178,14 +195,16 @@ def __init__(
...
-        elif self.loss == "trajectorybalance":
-            return list(self.forward_policy.model.parameters()) + list(
-                self.backward_policy.model.parameters()
-            )
-        else:
-            raise ValueError("Backward Policy cannot be a nn in flowmatch.")
+        parameters = list(self.forward_policy.model.parameters())
+        if self.backward_policy.is_model:
+            if self.loss == "flowmatch":
+                raise ValueError("Backward Policy cannot be a nn in flowmatch.")
```
Comment (alexhernandezgarcia) [inline gflownet/gflownet.py:201]:
```suggestion
                raise ValueError("Backward Policy cannot be a model in flowmatch.")
```

File: gflownet/gflownet.py [hunk 4]
```diff
@@ -178,14 +195,16 @@ def __init__(
...
+            parameters += list(self.backward_policy.model.parameters())
+        if self.state_flow is not None:
+            if self.loss != "forwardlooking":
+                raise ValueError(f"State flow cannot be trained in {self.loss} loss.")
```
Comment (alexhernandezgarcia) [inline gflownet/gflownet.py:205]:
```suggestion
                raise ValueError(f"State flow cannot be trained with {self.loss} loss.")
```

File: gflownet/utils/batch.py [hunk 5]
```diff
@@ -536,18 +551,26 @@ def _compute_parents(self):
...
         # Iterate over the trajectories to obtain the parents from the states
         for traj_idx, batch_indices in self.trajectories.items():
             # parent is source
             self.parents.append(self.envs[traj_idx].source)
+            # there's no source state in the batch
+            self.parents_indices.append(-1)
             # parent is not source
             # TODO: check if tensor and sort without iter
             self.parents.extend([self.states[idx] for idx in batch_indices[:-1]])
+            self.parents_indices.extend([idx for idx in batch_indices[:-1]])
```
Comment (alexhernandezgarcia) [inline gflownet/utils/batch.py:565]: No need to iterate over the indices in this case:
```suggestion
            self.parents_indices.extend(batch_indices[:-1])
```

File: gflownet/utils/batch.py [hunk 6]
```diff
@@ -521,6 +526,16 @@ def get_parents(
         else:
             return self.parents
 
+    def get_parents_indices(self):
```
Comment (alexhernandezgarcia) [inline gflownet/utils/batch.py:529]: Could you add docstring, please?

File: gflownet/utils/batch.py [hunk 7]
```diff
@@ -521,6 +526,16 @@ def get_parents(
...
+        if self.parents_available is False:
+            self._compute_parents()
+        return self.parents_indices
+
+    def get_parent_is_source(self):
```
Comment (alexhernandezgarcia) [inline gflownet/utils/batch.py:534]: Could you add docstring, please?

File: gflownet/utils/batch.py [hunk 8]
```diff
@@ -836,22 +859,66 @@ def get_rewards(
...
+
+    def _compute_rewards_parents(self):
+        """
+        Computes rewards of the self.parents by reusing rewards of the states (i.e. self.rewards).
+        Stores the result in self.rewards_parents
+        """
+        state_rewards = self.get_rewards()
+        self.rewards_parents = torch.zeros_like(state_rewards)
+        parent_is_source = self.get_parent_is_source()
+        parent_indices = self.get_parents_indices()
```
Comment (alexhernandezgarcia) [inline gflownet/utils/batch.py:892]: This is what I meant in https://github.com/alexhernandezgarcia/gflownet/pull/253/files#r1406451240:
```suggestion
        parent_indices = self.get_parents_indices()
        parent_is_source = parent_indices == -1
```

File: gflownet/gflownet.py (hunk 1 above)
Comment (carriepl) [reply gflownet/gflownet.py:753]: Oh yeah. Then it's worth updating the comment since it looks wrong even though it's right

File: gflownet/utils/batch.py [hunk 9]
```diff
@@ -521,9 +523,19 @@ def get_parents(
         else:
             return self.parents
 
+    def get_parents_indices(self):
+        """
+        Returns indices of the parents of the states in the batch.
+        Each index corresponds to the position of the patent in the self.states tensor, if it is peresent there.
```
Comment (alexhernandezgarcia) [inline gflownet/utils/batch.py:529]:
```suggestion
        Each index corresponds to the position of the parent in the self.states tensor, if it is present there.
```

File: gflownet/envs/base.py [hunk 10]
```diff
@@ -780,13 +780,13 @@ def traj2readable(self, traj=None):
...
 
-    def reward(self, state=None, done=None):
+    def reward(self, state=None, done=None, do_non_terminating=False):
         """
         Computes the reward of a state
         """
         state = self._get_state(state)
         done = self._get_done(done)
-        if done is False:
+        if done is False and do_non_terminating is False:
```
Comment (alexhernandezgarcia) [inline gflownet/envs/base.py:789]:
```suggestion
        if not done and not do_non_terminating:
```

File: tests/gflownet/utils/test_batch.py [hunk 11]
```diff
@@ -1325,3 +1330,167 @@ def test__make_indices_consecutive__multiplied_indices_become_consecutive(
...
+    rewards = []
+
+    # Iterate until envs is empty
+    while envs:
+        actions_iter = []
+        valids_iter = []
+        # Make step env by env (different to GFN Agent) to have full control
+        for env in envs:
+            parent = copy(env.state)
+            done_parent = env.done
```
Comment (alexhernandezgarcia) [inline tests/gflownet/utils/test_batch.py:1465]: Will a parent ever be done? :thinking:

Comment (AlexandraVolokhova) [reply tests/gflownet/utils/test_batch.py:1465]: hmm, should never be :) but I think it is more consistent to take this value from the env than hardcode it as False, no?

File: gflownet/utils/batch.py [hunk 12]
```diff
@@ -828,30 +845,85 @@ def get_rewards(
...
 
+    def get_rewards_parents(self) -> TensorType["n_states"]:
+        """
+        Returns the rewards of all parents in the batch
+        """
+        if not self.rewards_parents_available:
+            self._compute_rewards_parents()
+        return self.rewards_parents
+
+    def _compute_rewards_parents(self):
```
[the remark this answers is not in the document]
Comment (AlexandraVolokhova) [reply gflownet/utils/batch.py:911]: I've added several tests for` get_rewards_parents`, would you like me to add more for `_compute_rewards_parents` ?

File: tests/gflownet/utils/test_batch.py (hunk 11 above)
Comment (alexhernandezgarcia) [reply tests/gflownet/utils/test_batch.py:1465]: assert?

File: gflownet/utils/batch.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (alexhernandezgarcia) [reply gflownet/utils/batch.py:911]: No, that's enough

File: tests/gflownet/utils/test_batch.py (hunk 11 above)
Comment (alexhernandezgarcia) [reply tests/gflownet/utils/test_batch.py]: done in 0007bcc4314a3c6980acbb5cacc2af28be753142

**Review by alexhernandezgarcia — APPROVED** [review]
I think this is ready to merge!
